source/shared/cpp/ObjectModel/TextBlockText.cpp

```diff
@@ -0,0 +1,243 @@
+    }
+}
+
+void TextBlockText::AddTextSection(std::string text, std::string originalText, TextSectionFormat format)
+{
+    m_fullString.emplace_back(std::make_shared<TextSection>(text, originalText, format));
+}
+
+std::string TextBlockText::Concatenate()
+{
+    std::string formedString;
+    for (const auto& piece : m_fullString)
+    {
+        formedString += piece->GetOriginalText();
+    }
+    return formedString;
+}
+
+bool TextBlockText::IsValidTimeAndDate(const struct tm &parsedTm, int hours, int minutes)
+{
+    if (parsedTm.tm_mon <= 12 && parsedTm.tm_mday <= 31 && parsedTm.tm_hour <= 24 &&
+        parsedTm.tm_min <= 60 && parsedTm.tm_sec <= 60 && hours <= 24 && minutes <= 60)
+    {
+        if (parsedTm.tm_mon == 4 || parsedTm.tm_mon == 6 || parsedTm.tm_mon == 9 || parsedTm.tm_mon == 11)
+        {
+            return parsedTm.tm_mday <= 30;
+        }
+        else if (parsedTm.tm_mon == 2)
+        {
+            /// check for leap year
+            if ((parsedTm.tm_year % 4 == 0 && parsedTm.tm_year % 100 != 0) || parsedTm.tm_year % 400 == 0)
+            {
+                return parsedTm.tm_mday <= 29;
+            }
+
+            return parsedTm.tm_mday <= 28;
+        }
+
+        return true;
+    }
+    return false;
+}
+
+void TextBlockText::ParseDateTime(std::string in)
+{
+    const char dateDelimiter = '/';
+    const char timeDelimiter = ':';
+
+    std::vector<TextSection> sections;
+
+    std::regex pattern("\\{\\{((DATE)|(TIME))\\((\\d{4})-{1}(\\d{2})-{1}(\\d{2})T(\\d{2}):{1}(\\d{2}):{1}(\\d{2})(Z|(([+-])(\\d{2}):{1}(\\d{2})))((((, ?SHORT)|(, ?LONG))|(, ?COMPACT))|)\\)\\}\\}");
+    std::smatch matches;
+    std::string text = in;
+    enum MatchIndex
+    {
+        IsDate = 2,
+        Year = 4,
+        Month,
+        Day,
+        Hour,
+        Min,
+        Sec,
+        TimeZone = 12,
+        TZHr,
+        TZMn,
+        Format,
+        Style,
+    };
+    std::vector<int> indexer = {Year, Month, Day, Hour, Min, Sec, TZHr, TZMn};
+
+    while (std::regex_search(text, matches, pattern))
+    {
+        time_t offset{};
+        int  formatStyle{};
+        // Date is matched
+        bool isDate = matches[IsDate].matched;
+        int hours{}, minutes{};
+        struct tm parsedTm{};
+        int *addrs[] = {&parsedTm.tm_year, &parsedTm.tm_mon,
+            &parsedTm.tm_mday, &parsedTm.tm_hour, &parsedTm.tm_min,
+            &parsedTm.tm_sec, &hours, &minutes};
+
+        if (matches[Style].matched)
+        {
+            // match for long/short/compact
+            bool formatHasSpace = matches[Format].str()[1] == ' ';
+            int formatStartIndex = formatHasSpace ? 2 : 1;
+            formatStyle = matches[Format].str()[formatStartIndex];
+        }
+
+        AddTextSection(matches.prefix().str(), TextSectionFormat::RegularString);
+
+        if (!isDate && formatStyle)
+        {
+            AddTextSection(matches[0].str(), TextSectionFormat::RegularString);
```

Comment:
Is this needed to be saved?

Reply:
we used to append it to the output, so I guess it must be saved